### ISSUE TITLE
EPIC Parallel Search and Training

### ISSUE THREAD

**Issue by kmax12** [alteryx/evalml#144]
How can we take advantage of a cluster of machines to speed up search for optimal estimator

**Comment by christopherbunn** [comment]
After talking with PS, we have the green light to reopen this issue and move forward with this. For now, the implementation will be limited to a local implementation, but ideally the changes that we make to support this should transfer over nicely to a distributed version. A scoped-down version of this plan is available [here](https://alteryx.quip.com/mz5SAyv0D9pF/Single-Machine-Parallel-EvalML).

**Comment by christopherbunn** [comment]
Implementation Plan 

Phase 1: parallel automl search (3 weeks 3 days)

* Show automl batching to users (3 days)
    * Move batching as a public parameter for AutoMLSearch
    * Update the search progress log output for batching
    * Update docs to reflect batching and how that works
* Refactor automl search (3 days)
    * Define an evaluation method (name can be whatever) which takes pipeline+data input, outputs scores/results/CV data etc., with no side-effects
    * To do this: update _evaluate to call _add_result afterwards instead of inside that method, so that _evaluate has no side effects.
    * May have to update _compute_cv_scores too
* Add API for “search engine”. Initial implementation preserves existing sequential evaluation. (1 week)
    * Create structures for engines 
    * Implement base class + SequentialEngine
    * Add in docs for defining your own execution engine
    * Throughly unit + perf test implementation to ensure consistency
* Add in Dask engine (1 week)
    * Implement local support
    * Unit test + get scalability metrics
    * Introduce this as method of officially supported distributed 
* Documentation: make sure API ref is tight. Add user guide for how to run automl with sequential vs parallel. (2 days)

Phase 2: testing (4 weeks)
* Performance testing adjustments and measuring (2 week)
    * Evaluate on resources available, 
    * Refine implementation as necessary
    * Test on different environments
        * Linux and windows
        * Machines with different resources
            * Not only beefier VMs on AWS, but also standard machines with less resources (2 core VMs, etc.) 
        * Default vs. common parameter combinations
        * Edge cases
    * Test with 4 cores (existing looking_glass, ECS)
    * Test with more than 4 cores (would require time for looking glass updates, to use EC2-backed ECS, or to use EKS)
* Implementation tweaking and further performance testing (2 week)
    * Refinement of resource checking mechanism
    * Plan out extension project for distributed

*Key Dates*
Oct 27th is the Oct milestone release. Nov 24th is Nov release.

*Goal*
Finish phase 1 (impl) by Fri Nov 6th. Then do testing. Include this feature in Nov release.

*Stretch Goal*
Finish phase 1 (impl) by Oct 27th, for the Oct release.

**Comment by dsherry** [comment]
We currently have `n_jobs` defaulting to `-1` for estimators which support that parameter (sklearn), meaning all CPUs will be used during training. The question is, why? 😂

[Sklearn's doc says](https://scikit-learn.org/stable/glossary.html#term-n-jobs):
> This parameter is used to specify how many concurrent processes or threads should be used for routines that are parallelized with joblib.
>
> n_jobs is an integer, specifying the maximum number of concurrently running workers. If 1 is given, no joblib parallelism is used at all, which is useful for debugging. If set to -1, all CPUs are used. For n_jobs below -1, (n_cpus + 1 + n_jobs) are used. For example with n_jobs=-2, all CPUs but one are used.
>
> n_jobs is None by default, which means unset; it will generally be interpreted as n_jobs=1, unless the current joblib.Parallel backend context specifies otherwise.
> 
> For more details on the use of joblib and its interactions with scikit-learn, please refer to our parallelism notes.

@christopherbunn @jeremyliweishih I know we already talked about this. We should keep this in mind as we move forward. I think we should update the default `n_jobs` to `None` as part of this project.

**Comment by dsherry** [comment]
Update from discussion with @christopherbunn @jeremyliweishih @rpeck:

@christopherbunn has made great progress on this! Shared some initial local performance testing results on tuning the number of parallel workers across several automl batches.

**Next steps**
* Plan:
    * Merge engine API and sequential impl by Weds Dec 9th (still need to figure out the search iteration plot part)
    * Then merge parallel impl by Dec 18th
* @christopherbunn is gonna share a link to the ipython notebooks he's got at the moment.

**Comment by christopherbunn** [comment]
Here's the [notebook](https://alteryx0-my.sharepoint.com/:u:/g/personal/chris_bunn_alteryx_com/EeEW33t_fgdLkssGddV0lPcBPryxuxWVdZLRt4e1bUwEtw?e=QbHc3j) with a basic workflow for the Dask engine. You'll need to change to the `1297_dask_engines` branch to use it.

**Comment by christopherbunn** [comment]
Currently, there are two outstanding PRs that are related to this epic. #1433 should be merged in before #1525.

#1433 (Engine API) Notes:
- What's left to do:
  - Additional test cases for the standalone engine. While an "engine" is a pretty integral part of the search process (and should be covered by AutoML Search test cases), there still needs to be some more direct testing.  

#1525 (Dask Engine) Notes:
-  What's left to do:
   - Fleshing out more testing for Dask engine. I have a handful already that should address basic functionality but it could be improved. 
   - Developing a way to measure/test for parallel performance. Ideally it should work in the existing EvalML looking glass system but I could see it being a separate script until the necessary changes are made. Happy to discuss more about metrics, design, etc.  

While it's out of scope for this epic, a few directions that we can take once the top two PRs get merged in:
- Distributed evaluation of pipelines. It's less additional implementation (since Dask should be able to take care of this) and more testing/developing documentation. All you should need to do is [setup a remote Dask cluster/scheduler](https://docs.dask.org/en/latest/setup.html), create a local `DaskClient()` that points to  
- Evaluate pipelines asynchronously. It might reduce the amount of idle workers and should speed up searches. This would involve breaking out of the current batch system that we have, so definitely a bit further out in the future. 
- GPU support. There's a number of ways that we could do this, going from having a preprocessing step that would enable the relevant flags (like in [XGBoost](https://xgboost.readthedocs.io/en/latest/gpu/)) all the way to having the engine dump the parameter pipelines into GPU version (using something like [RAPIDS](https://docs.rapids.ai/api/cuml/stable/cuml_intro.html#where-possible-match-the-scikit-learn-api)). It could potentially build on top of the Dask engine or be it's own separate engine all together.

**Comment by dsherry** [comment]
Synced with @christopherbunn . #1433 (engine API PR) is almost ready for a second round of review. Goal is to get that up for review today and to merge that this week.

The next outstanding part is the dask engine implementation. That's currently a draft, #1525. Goal is to complete that by the end of the next sprint, which is the 26th.
